src/vm/typedesc.cpp

```diff
@@ -363,106 +364,111 @@ BOOL TypeDesc::CanCastTo(TypeHandle toType, TypeHandlePairList *pVisited)
     }
     CONTRACTL_END
 
-    if (TypeHandle(this) == toType)
+    if (TypeHandle(this) == toTypeHnd)
         return TRUE;
 
-    //A boxed variable type can be cast to any of its constraints, or object, if none are specified
-    if (IsGenericVariable())
-    {
-        TypeVarTypeDesc *tyvar = (TypeVarTypeDesc*) this;
-
-        DWORD numConstraints;
-        TypeHandle *constraints = tyvar->GetConstraints(&numConstraints, CLASS_DEPENDENCIES_LOADED);
+    BOOL fCast = FALSE;
 
-        if (toType == g_pObjectClass)
-            return TRUE;
+    if (IsArray())
+    {
+        MethodTable* pMT = this->GetMethodTable();
```

Comment:
This refactoring of array casts does not look right. This:
```
using System;

interface IFace
{
}

class G<T> where T:class,IFace
{
}

class My
{
    static void Main()
    {
        var a = typeof(G<>).GetGenericArguments()[0].MakeArrayType();
        Console.WriteLine(typeof(IFace[]).IsAssignableFrom(a));
    }
}
```
will print true before your change, but false after your change. (Add a test...)

Reply:
Hmm. `GetMethodTable` lies here. Returns `object[]` for `T[]`.

Reply:
Right, GetMethodTable() is not always representative for arrays.

Reply:
I thought the only remaining case was when conversion code was manufacturing temporary typedescs.
Is this one temporary too?

Reply:
Aaaargh

```C
    // Strictly speaking no method table should be needed for
    // arrays of the faked up TypeDescs for variable types that are
    // used when verfifying generic code.
    // However verification is tied in with some codegen in the JITs, so give these
    // the shared MT just in case.
    // This checks match precisely one in ParamTypeDesc::OwnsMethodTable
    if (CorTypeInfo::IsGenericVariable(elemType)) {
        // This is loading the canonical version of the array so we can override
        OVERRIDE_TYPE_LOAD_LEVEL_LIMIT(CLASS_LOADED);
        return(ClassLoader::LoadArrayTypeThrowing(TypeHandle(g_pObjectClass), arrayKind, Rank).GetMethodTable());
    }
```